crates/resolver/src/resolver.rs

```diff
@@ -122,39 +195,7 @@ impl<R: ConnectionProvider> Resolver<R> {
     /// * `options` - basic lookup options for the resolver
     /// * `provider` - connection provider, for DNS connections, I/O, and timers
     pub fn new(config: ResolverConfig, options: ResolverOpts, provider: R) -> Self {
-        let pool = NameServerPool::from_config_with_provider(&config, options.clone(), provider);
-        let either;
-        let client = RetryDnsHandle::new(pool, options.attempts);
-        if options.validate {
-            #[cfg(feature = "__dnssec")]
-            {
-                use crate::proto::dnssec::DnssecDnsHandle;
-                either = LookupEither::Secure(DnssecDnsHandle::new(client));
-            }
-
-            #[cfg(not(feature = "__dnssec"))]
-            {
-                // TODO: should this just be a panic, or a pinned error?
-                tracing::warn!("validate option is only available with 'dnssec' feature");
-                either = LookupEither::Retry(client);
-            }
-        } else {
-            either = LookupEither::Retry(client);
-        }
-
-        let hosts = match options.use_hosts_file {
-            ResolveHosts::Always | ResolveHosts::Auto => Some(Arc::new(Hosts::new())),
-            ResolveHosts::Never => None,
-        };
-
-        trace!("handle passed back");
-        let lru = DnsLru::new(options.cache_size, dns_lru::TtlConfig::from_opts(&options));
-        Self {
-            config,
-            client_cache: CachingClient::with_cache(lru, either, options.preserve_intermediates),
-            options,
-            hosts,
-        }
+        Self::builder(config, options, provider).build()
```

Comment:
Want to advocate for making this more modular:

```rust
impl<R> Resolver<R> {
    // Currently `from_system_conf()`
    fn builder(provider: R) -> ResolverBuilder<R> { .. }
    fn builder_with_config(config: ResolverConfig, provider: R) -> ResolverBuilder<R> { .. }
    ..
}
```

The `ResolverBuilder` will store `ResolverOpts::default()` and expose an `options_mut(&mut self) -> &mut ResolverOpts` method. Drop `new()` in favor of letting callers go through the builder.